cluster-autoscaler/cloudprovider/gce/gce_manager.go

```diff
@@ -661,15 +661,43 @@ func (m *gceManagerImpl) fetchExplicitMigs(specs []string) error {
 	return nil
 }
 
-func (m *gceManagerImpl) buildMigFromSpec(spec string) (*Mig, error) {
-	s, err := dynamic.SpecFromString(spec, scaleToZeroSupported)
+func (m *gceManagerImpl) buildMigFromFlag(flag string) (Mig, error) {
+	s, err := dynamic.SpecFromString(flag, scaleToZeroSupported)
 	if err != nil {
 		return nil, fmt.Errorf("failed to parse node group spec: %v", err)
 	}
-	mig := &Mig{gceManager: m, minSize: s.MinSize, maxSize: s.MaxSize, exist: true}
-	if mig.Project, mig.Zone, mig.Name, err = ParseMigUrl(s.Name); err != nil {
+	return m.buildMigFromSpec(s)
+}
+
+func (m *gceManagerImpl) buildMigFromAutoCfg(link string, cfg cloudprovider.MIGAutoDiscoveryConfig) (Mig, error) {
+	s := &dynamic.NodeGroupSpec{
+		Name:               link,
+		MinSize:            cfg.MinSize,
+		MaxSize:            cfg.MaxSize,
+		SupportScaleToZero: scaleToZeroSupported,
+	}
+	return m.buildMigFromSpec(s)
+}
+
+func (m *gceManagerImpl) buildMigFromSpec(s *dynamic.NodeGroupSpec) (Mig, error) {
```

Comment:
Extracting this is a drive-by cleanup, that has nothing to do with this refactor. I'm always somewhat worried about those, especially in cases like that, where we're changing code specific to feature we don't test at all (autodiscovery). I think that just including this means we should probably include autodiscovery in your post-refactor testing.

Regardless of the above this should at least be a separate PR. Reviewing just this change would be quite simple, but it adds up, resulting in this huge PR (which it turn makes me delay review until I have enough time to go through the whole thing, rather than just keep reviewing as you send PRs).

Reply:
Fair point. Although more PRs also mean more review cycles :)